### PR TITLE
Policy engine core

### PR DESCRIPTION
**Pull request by vadim-kovalyov**
**Context:**
There are many use cases in the IoT Platform (IoT Edge and IoT Hub) that require some sort of policy engine to evaluate policies against certain input and state of the system. The main example is authorization for MQTT Broker.  

**This PR:**
Adds generic json-based authorization policy engine. Policy uses simple Identity->Operation->Resource approach to define authorization rules. Below is an example of a policy definition consisting of two rules.

```
{
            "schemaVersion": "2020-10-30",
            "statements": [
                {
                    "effect": "deny",
                    "identities": [
                        "actor_a"
                    ],
                    "operations": [
                        "write"
                    ],
                    "resources": [
                        "resource_1"
                    ]
                },
                {
                    "effect": "allow",
                    "identities": [
                        "actor_b"
                    ],
                    "operations": [
                        "read"
                    ],
                    "resources": [
                        "resource_1"
                    ]
                }
            ]
        }
```

**Open questions:**
- Exact location of the policy core code.

**Next steps:**
- Integration w/ MQTT Broker. Broker specific substitutor, validator, resource marcher implementations. (separate PRs)

### CODE REVIEW

File: mqtt/policy/src/core/builder.rs [hunk 1]
```diff
@@ -0,0 +1,706 @@
...
+    lazy_static! {
+        static ref VAR_PATTERN: Regex =
+            Regex::new(r#"\{\{[^\{\}]+\}\}"#).expect("failed to create a Regex from pattern");
+    }
+    VAR_PATTERN.is_match(value)
+}
+
+#[derive(Deserialize)]
+#[serde(rename_all = "camelCase")]
+struct PolicyDefinition20201030 {
```
Comment (dylanbronson) [inline mqtt/policy/src/core/builder.rs:249]: How often do we see this naming scheme changing? 
Are Statement/Effect/PolicyDefinition tied together?

File: mqtt/policy/src/core/builder.rs [hunk 2]
```diff
@@ -0,0 +1,706 @@
...
+    fn into(self) -> EffectOrd {
+        match self.effect {
+            Effect20201030::Allow => EffectOrd::new(CoreEffect::Allow, self.order),
+            Effect20201030::Deny => EffectOrd::new(CoreEffect::Deny, self.order),
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
```
Comment (dylanbronson) [inline mqtt/policy/src/core/builder.rs:284]: Why not put the tests in a separate file?

Comment (vadim-kovalyov) [reply mqtt/policy/src/core/builder.rs:284]: it is rust practice to keep unit tests in the same file, as a submodule.

You can add tests in a separate file - but those will be integration tests for the crate.

File: mqtt/policy/src/core/builder.rs (hunk 1 above)
Comment (vadim-kovalyov) [reply mqtt/policy/src/core/builder.rs:249]: >How often do we see this naming scheme updating?

I'm not sure how often that would be, but we need to think about versioning.

>Are Statement/Effect/PolicyDefinition tied together?

yes, in current policy definition schema.

File: mqtt/policy/src/core/mod.rs [hunk 3]
```diff
@@ -0,0 +1,576 @@
...
+    }
+}
+
+impl From<BTreeMap<String, EffectOrd>> for Resources {
+    fn from(map: BTreeMap<String, EffectOrd>) -> Self {
+        Resources(map)
+    }
+}
+
+/// Represents a request that needs to be `evaluat`ed by `Policy` engine.
```
Comment (dylanbronson) [inline mqtt/policy/src/core/mod.rs:224]: nit: tick mark

File: mqtt/policy/src/errors.rs [hunk 4]
```diff
@@ -0,0 +1,21 @@
+use thiserror::Error;
+
+/// A specialized `Result` type for mini-redis operations.
```
Comment (dmolokanov) [inline mqtt/policy/src/errors.rs:3]: _mini-redis_  :thinking:

File: mqtt/policy/Cargo.lock [hunk 5]
```diff
@@ -0,0 +1,165 @@
+# This file is automatically @generated by Cargo.
```
Comment (dmolokanov) [inline mqtt/policy/Cargo.lock:1]: I think it should be removed from the workspace

**Review by dmolokanov — APPROVED** [review]
Great work! I like the way code is structure. Super neat!

File: mqtt/policy/src/errors.rs (hunk 4 above)
Comment (vadim-kovalyov) [reply mqtt/policy/src/errors.rs:3]: Oops.. ^-^

**Review by dmolokanov — DISMISSED** [review]
:shipit:

**Review by dmolokanov — APPROVED** [review]
🚀

**Pull request merged**